codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java

```diff
@@ -190,6 +210,121 @@ public static void writeFrameHeader(ByteBuf out, int payloadLength, byte type,
+        private final ChannelPromise promise;
+        private int expectedCount;
+        private int successfullCount;
+        private int failureCount;
+        private boolean doneAllocating;
+
+        SimpleChannelPromiseAggregator(ChannelPromise promise, Channel c) {
+            super(c);
+            assert promise != null;
+            this.promise = promise;
+        }
+
+        SimpleChannelPromiseAggregator(ChannelPromise promise, Channel c, EventExecutor e) {
+            super(c, e);
+            assert promise != null;
+            this.promise = promise;
+        }
+
+        /**
+         * Allocate a new promise which will be used to aggregate the overall success of this promise aggregator.
+         * @return A new promise which will be aggregated.
+         * {@code null} if {@link #doneAllocatingPromises()} was previously called.
+         */
+        public ChannelPromise newPromise() {
+            if (doneAllocating) {
+                return null;
+            }
+            ++expectedCount;
+            return this;
+        }
+
+        /**
+         * Signify that no more {@link #newPromise()} allocations will be made.
+         * The aggregation can not be successful until this method is called.
+         * @return The promise that is the aggregation of all promises allocated with {@link #newPromise()}.
+         */
+        public ChannelPromise doneAllocatingPromises() {
+            doneAllocating = true;
+            if (successfullCount == expectedCount) {
+                promise.setSuccess();
+                return super.setSuccess();
+            }
+            return this;
+        }
+
+        @Override
+        public boolean tryFailure(Throwable cause) {
+            if (allowNotificationEvent()) {
+                ++failureCount;
+                if (failureCount == 1) {
+                    promise.tryFailure(cause);
+                    return super.tryFailure(cause);
+                }
+                // TODO: We break the interface a bit here.
+                // Multiple failure events can be processed without issue because this is an aggregation.
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * Fail this object if it has not already been failed.
+         * <p>
+         * This method will NOT throw an {@link IllegalStateException} if called multiple times
+         * because that may be expected.
+         */
+        @Override
+        public ChannelPromise setFailure(Throwable cause) {
+            if (allowNotificationEvent()) {
+                ++failureCount;
+                if (failureCount == 1) {
+                    promise.setFailure(cause);
+                    return super.setFailure(cause);
+                }
+            }
+            return this;
+        }
+
+        private boolean allowNotificationEvent() {
+            return successfullCount + failureCount < expectedCount;
```

Comment:
Perhaps I need more details.  I'm thinking it is beneficial to notify as soon as we can, for example in the case of failure we can short circuit waiting for all other events to complete.